### PR TITLE
refactor(snack-runtime): split runtime app into reusable package and app

### PR DESCRIPTION
**Pull request by byCedric**
# Why

This allows some of our partners to reuse the Snack infrastructure to run Snacks.

# How

- [x] Copied the current `/runtime` code to `/packages/snack-runtime`
- [x] Moved the required changes from the [custom runtime fork](https://github.com/byCedric/snack/tree/%40bycedric/snack-runtime-46)
- [x] Updated our CI for `snack-runtime`
- [ ] ~~Cleaned up the old `/runtime` app to use the `/packages/snack-runtime` app~~ → Postponing in a separate PR
- [x] Published and replaced all `file:../` references in dependencies
- [x] Added context-based API to set the aliased modules for `snack-runtime`

# Test Plan

TBD

### CODE REVIEW

File: runtime-shell/package.json [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+  "main": "index.js",
+  "owner": "exponent",
+  "scripts": {
+    "start": "expo start",
+    "web": "expo start --web",
+    "postinstall": "patch-package",
+    "lint": "eslint ./src ./types",
+    "typescript": "tsc",
+    "test": "jest",
+    "deploy:staging": "EXPO_STAGING=1 expo-cli publish --clear",
```
Comment (Kudo) [inline runtime-shell/package.json:16]:
```suggestion
    "deploy:staging": "CLOUD_ENV=staging EXPO_STAGING=1 expo-cli publish --clear",
```

File: runtime-shell/package.json [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    "deploy:prod": "CLOUD_ENV=production NODE_ENV=production expo-cli publish --clear",
+    "deploy:web:staging": "node ./web/deploy-script.js",
```
Comment (Kudo) [inline runtime-shell/package.json:18]:
```suggestion
    "deploy:web:staging": "CLOUD_ENV=staging node ./web/deploy-script.js",
```

File: packages/snack-runtime/src/transports/RuntimeTransportImplSocketIO.ts [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+  private readonly sender: string;
+  private readonly socket: Socket<ServerToClientEvents, ClientToServerEvents>;
+
+  constructor(device: Device) {
+    this.device = device;
+    this.sender = JSON.stringify(device);
+    const snackpubURL =
+      Constants.manifest?.extra?.cloudEnv !== 'production'
+        ? SNACKPUB_URL_STAGING
+        : SNACKPUB_URL_PRODUCTION;
```
Comment (Kudo) [inline packages/snack-runtime/src/transports/RuntimeTransportImplSocketIO.ts:36]:
```suggestion
      Constants.manifest?.extra?.cloudEnv === 'staging'
        ? SNACKPUB_URL_STAGING
        : SNACKPUB_URL_PRODUCTION;
```

File: packages/snack-runtime/src/Modules.tsx [hunk 4]
```diff
@@ -0,0 +1,902 @@
...
+            'Loaded dependency',
+            cacheHandle,
+            `from cache ${bundle ? bundle.length : undefined} bytes`
+          );
+        } else {
+          // In development, try fetching from staging cloudfront first
+          const cloudFrontUrls =
+            Constants.manifest?.extra?.cloudEnv !== 'production'
+              ? [SNACKAGER_CDN_STAGING, SNACKAGER_CDN_PROD]
+              : [SNACKAGER_CDN_PROD];
```
Comment (Kudo) [inline packages/snack-runtime/src/Modules.tsx:318]:
```suggestion
            Constants.manifest?.extra?.cloudEnv === 'staging'
              ? [SNACKAGER_CDN_STAGING, SNACKAGER_CDN_PROD]
              : [SNACKAGER_CDN_PROD];
```

File: packages/snack-runtime/src/App.tsx [hunk 5]
```diff
@@ -0,0 +1,623 @@
...
+      clearTimeout(this._askTimeout);
+      this._askTimeout = undefined;
+    }
+  };
+
+  _uploadPreviewToS3 = async (asset: string, height: number, width: number) => {
+    const url = `${
+      Constants.manifest?.extra?.cloudEnv !== 'production'
+        ? API_SERVER_URL_STAGING
+        : API_SERVER_URL_PROD
```
Comment (Kudo) [inline packages/snack-runtime/src/App.tsx:400]:
```suggestion
      Constants.manifest?.extra?.cloudEnv === 'staging'
        ? API_SERVER_URL_STAGING
        : API_SERVER_URL_PROD
```

**Review by Kudo — CHANGES_REQUESTED** [review]
mostly looks good and i like the `runtime-shell` pretty much. that's a clever solution.

just leave one problem from https://github.com/byCedric/snack/pull/5 where partners' snack-runtimes does not have the `Constants.manifest.extra.cloudEnv`, they will reach to staging servers. we maybe change the logic to production-first and let runtime-shell to have staging support.

File: runtime-shell/package.json (hunk 1 above)
Comment (byCedric) [reply runtime-shell/package.json:16]: Thanks for catching that!

**Review by Kudo — APPROVED** [review]
🚀🚀🚀 excited for this

**Comment by byCedric** [comment]
Tested this in https://github.com/byCedric/snack-custom-runtime, and it seems good to go!

**Pull request merged**